pkg/operator/ceph/controller/network.go

```diff
@@ -0,0 +1,419 @@
+	ranges []string,
+	err error,
+) {
+	ranges = []string{}
+	clusterNamespace := clusterInfo.Namespace
+
+	netSelection, err := clusterInfo.NetworkSpec.GetNetworkSelection(clusterNamespace, cephNetwork)
+	if err != nil {
+		return ranges, errors.Wrapf(err, "failed to get %q network selection", cephNetwork)
+	}
+	if netSelection == nil {
+		// no selection for this network
+		return ranges, nil
+	}
+
+	// Modify the user-provided network selection to set the interface name in the pod be the Ceph
+	// network (e.g., public, cluster). This is important for accurately identifying the CIDR range
+	// provided by the network selection in the canary. It is safe to make this change to the
+	// user-defined spec in the canary for this limited purpose, but not safe during Ceph runtime.
+	netSelection.InterfaceRequest = string(cephNetwork)
+
+	netSelectionValue, err := cephv1.NetworkSelectionsToAnnotationValue(netSelection)
+	if err != nil {
+		return ranges, errors.Wrapf(err, "failed to get %q network annotation value for cluster in namespace %q", cephNetwork, clusterNamespace)
+	}
+
+	logger.Debugf("discovering ceph %q network CIDR(s) for cluster %q", cephNetwork, clusterNamespace)
+
+	// This job is complex to overcome 2 specific limitations:
+	// - CNI/Multus' `k8s.v1.cni.cncf.io/network-status` annotation only reports IP addrs, not CIDRs
+	// - `ip addr show` in the pod shows CIDRs but also contains extra IPv6 SLAAC addrs in addition
+	//   to the addr(s) attached by CNI/Multus
+	// Rook must cross-reference both pieces of info to accurately read the CIDR(s) for the net.
+	// Both pieces of info must come from the same Pod+Container in order to be cross-ref'd.
+	// Use downward API to allow the cmd reporter job to report both pieces of info at the same time
+	networkCanary, err := newCmdReporter(
+		clusterdContext.Clientset,
+		clusterInfo.OwnerInfo,
+		"rook-ceph-network-canary",
+		"rook-ceph-network-"+string(cephNetwork)+"-canary",
+		clusterNamespace,
+		[]string{"bash", "-c", `set -e
+			cat /tmp/network-status
+			echo "" # newline
+			echo "` + separator() + `"
+			ip --json address show dev ` + string(cephNetwork) + `
+			`},
+		[]string{},
+		rookImage,
+		clusterSpec.CephVersion.Image,
+		clusterSpec.CephVersion.ImagePullPolicy,
+	)
+	if err != nil {
+		return ranges, errors.Wrapf(err, "failed to set up ceph %q network canary", cephNetwork)
+	}
+
+	job := networkCanary.Job()
+	job.Spec.Template.Spec.ServiceAccountName = "rook-ceph-cmd-reporter"
+
+	// put command reporter job on the multus network
+	job.Spec.Template.Annotations = map[string]string{
+		nadv1.NetworkAttachmentAnnot: netSelectionValue,
+	}
+
+	// use osd placement for net canaries b/c osd pods are present on both public and cluster nets
+	clusterSpec.Placement[cephv1.KeyOSD].ApplyToPodSpec(&job.Spec.Template.Spec)
+
+	// TODO: do we need resource requests/requirements? it's just sleeping
+	// ... if needed, maybe logcollector, with small requirements would be a good option?
```

Comment:
@travisn thoughts?

Reply:
Some environments are strict about requiring resource requests/limits on all pods, so we at least need an option. The requirements are going to be very low. I'd suggest we define a new category of resources for this in the cluster CR, though by default I'd say we don't need to set them.

Reply:
I removed this todo without resolving it in this PR. It looks like the cmd reporter doesn't set any resource requests/requirements at all, and it looks like `DetectCephVersion()` also doesn't set any resources. Can we instead take an item to handle resources for the cmd-reporter as a general work item? I think it'll bloat this more, and it seems like we don't have anyone with any big problems around this today.

Reply:
Oh interesting, if the cmd reporter already doesn't set requests/requirements, then we can skip it for now.